pkg/urbit/jets/e/lms.c

```diff
@@ -0,0 +1,1492 @@
+/* jets/e/lms.c
+**
+::
+::::  Vector type in single-precision floating-point @rs
+  ::
+::  Conventions:
+::
+::  m,n,p   always dimensions
+::  i,j,k   always indices (also ii,jj,kk for coder-spec not user-spec)
+::  a,b,c   always lists
+::  u,v,w   always vector/matrix atoms
+::  s,t     always real/floats
+::  `_` suffix indicates a C array unpacking of an @lv atom
+*/
+#include "all.h"
+#include <softfloat.h>
+#include <stdio.h>
+#include <math.h>
+/*#include <atlas.h>*/
+
+#define max(a,b) \
+  ({ __typeof__ (a) _a = (a); \
+      __typeof__ (b) _b = (b); \
+    _a > _b ? _a : _b; })
+#define min(a,b) \
+  ({ __typeof__ (a) _a = (a); \
+      __typeof__ (b) _b = (b); \
+    _a < _b ? _a : _b; })
+
+#define SINGNAN 0x7fc00000
+
+  union trip {
+    float32_t s;    //struct containing v, uint_32
+    c3_w c;         //uint_32
+    float b;        //float_32, compiler-native, useful for debugging printfs
+  };
+
+/* functions
+*/
+  static inline c3_t
+  _nan_test(float32_t a)
+  {
+    return !f32_eq(a, a);
+  }
+
+  static inline float32_t
+  _nan_unify(float32_t a)
+  {
+    if ( _nan_test(a) )
+    {
+      *(c3_w*)(&a) = SINGNAN;
+    }
+    return a;
+  }
+
+  static inline void
+  _set_rounding(c3_w a)
+  {
+    switch ( a )
+    {
+    default:
+      u3m_bail(c3__fail);
+      break;
+    case c3__n:
+      softfloat_roundingMode = softfloat_round_near_even;
+      break;
+    case c3__z:
+      softfloat_roundingMode = softfloat_round_minMag;
+      break;
+    case c3__u:
+      softfloat_roundingMode = softfloat_round_max;
+      break;
+    case c3__d:
+      softfloat_roundingMode = softfloat_round_min;
+      break;
+    }
+  }
+
+  struct
+  dims
+  {
+    c3_w rows, cols;
+  };
+
+  struct dims
+  shape(u3_atom p)  /* @ud */
+  {
+    c3_w mn = u3r_met(3, p)/4 - 1;  // total elements
+    c3_w d = u3r_word(mn+1, p);     // dimensionality, for now presumptively 2
+    c3_w m = u3r_word(mn, p);       // number of rows
+
+    struct dims shp;
+    shp.rows = m;
+    shp.cols = (c3_w)(mn / m);              // number of columns
+
+    return shp;
+  }
+
+  void
+  print_matrix(c3_w* u,
+               uint32_t m_u,
+               uint32_t n_u)
+  {
+    c3_w ii, jj;
+    union trip a;
+    fprintf(stderr, "[%u x %u] matrix:\n\r", m_u, n_u);
+    for ( ii = 1; ii <= m_u; ii++ )
+    {
+      fprintf(stderr, "[ ");
+      for ( jj = 1; jj <= n_u; jj++ )
+      {
+        a.c = u[(ii-1)*n_u+jj-1];
+        fprintf(stderr, "%f ", a.b);
```

Comment:
This may be fine for tests - I'm failing to track down what's calling this actually - but for general use, note that printing rows individually can significantly disrupt vertical alignment. Compare:
![18513121-F28A-48A4-A2C0-7B9336E62848](https://user-images.githubusercontent.com/1857414/125209698-89601000-e24f-11eb-8f9e-c810aee6dd2a.jpeg)

Reply:
Yeah, it's terrible but it's basically a debug printf.  Not currently in use but frequently helpful.